### DISCUSSION TITLE
How to enable static linking for rnpgp library and its dependencies

### DISCUSSION THREAD

**Discussion by shemeshg** [rnpgp/rnp#2168] [Q&A] [answered]
Hi,

I have a question about **static** linking rnpgp library and its dependencies to my application.

 Using Linux, GCC Vcpkg setup. I have installed rnpgp library and its dependencies (json-c and botan-3) using vcpkg. However, when I try to deploy my application using qtdeploy, I find that the shared libraries of rnpgp and its dependencies are not copied to the binary folder. This causes my application to fail to run on other machines that do not have these libraries installed.

I want to enable static linking for rnpgp library and its dependencies, so that I can distribute my application as a single executable file. I have tried to modify the CMakeLists.txt file of rnpgp library to use static linking instead of INTERFACE_LINK_LIBRARIES, but I get errors during the compilation process.

Is this an issue? How can I enable static linking for rnpgp library and its dependencies using vcpkg and CMake? Do you have any examples or documentation that I can refer to?

I appreciate your help and support. 
Thanks for your time and attention.

## Logs and data
- Simple test project:
https://github.com/shemeshg/rnpgp-quickstart/blob/main/RnpCore/CMakeLists.txt

- already opened issue maybe related because vcpkg is not part of the `apt-get` dependency management. 
and no `sudo apt-get install rnpgp` yet
https://github.com/rnpgp/rnp/issues/1648

-  Error for qt app using rnpgp on fresh new Ubuntu 20.04 machine
```
ubuntu@ubuntutest:~/Downloads/pass-simple-0.145-Linux/opt/pass-simple/bin$ ./pass-simple 
./pass-simple: error while loading shared libraries: librnp.so.0: cannot open shared object file: No such file or directory
ubuntu@ubuntutest:~/Downloads/pass-simple-0.145-Linux/opt/pass-simple/bin$ 
```
- ldd after deploy:
```
...
...
libjson-c.so.5 => /lib/x86_64-linux-gnu/libjson-c.so.5 (0x00007f4556776000)
libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f4556763000)
libassuan.so.0 => /lib/x86_64-linux-gnu/libassuan.so.0 (0x00007f455674d000)
libbotan-3.so.2 => /lib/libbotan-3.so.2 (0x00007f4557d1b000)
...
```

- the shared linking commands at the `cmake`
```
pkg_check_modules(JSONC IMPORTED_TARGET json-c12)
if(NOT JSONC_FOUND)
    pkg_check_modules(JSONC REQUIRED IMPORTED_TARGET json-c)
endif(NOT JSONC_FOUND)

add_library(JSON-C::JSON-C INTERFACE IMPORTED)
set_target_properties(JSON-C::JSON-C PROPERTIES INTERFACE_LINK_LIBRARIES PkgConfig::JSONC)

pkg_check_modules(Botan REQUIRED IMPORTED_TARGET botan-3)
add_library(Botan::Botan  INTERFACE IMPORTED)
set_target_properties(Botan::Botan PROPERTIES INTERFACE_LINK_LIBRARIES PkgConfig::Botan)
```

**Comment by ni4** [comment, marked as answer]
RNP's CMake parameter `DBUILD_SHARED_LIBS=Off` should help here, If I'm not mistaken. Examples are available Github workflow files. Or does this inflict some problems? It's attempt to 'quick answer', if it doesn't help I'll look into a problem a bit later (or maybe @maxirmx could be willing to help here).

**Comment by shemeshg** [comment]
Thanks, `BUILD_SHARED_LIBS=OFF` it did solved the problem.

## log
```
ubuntu@ubuntu:~/Documents/pass-simple-qt/gpgfactory/build/main$ ldd /usr/bin/rnp
        linux-vdso.so.1 (0x00007ffe89fc7000)
        libjson-c.so.5 => /lib/x86_64-linux-gnu/libjson-c.so.5 (0x00007f1d21a89000)
        libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f1d21a76000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f1d21a5a000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f1d2182e000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f1d21747000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f1d21727000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f1d214fd000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f1d21eea000)
ubuntu@ubuntu:~/Documents/pass-simple-qt/gpgfactory/build/main$ 
```

the `json-c` at the log is ok.  I've checked against fresh 20.04 install, and `json-c` is part of the OS.

Thanks

**Comment by maxirmx** [comment]
@shemeshg 
BUILD_SHARED_LIBS=OFF will create ststic rnp library. We cannot affect selection of other libraries for your application through rnp cmake script.
If you need static json-c and static botan libraries you have to use appropriate vcpkg triplet and/or refer to static libraries it you build scripts.

As @ni4  noted, there are various build scenarious implemented and tested in GHA workflows albeight not exactly yours. You may find Windows action scrip useful since it is vcpkg based.

Thank you
